Update intra-doc link documentation to match the implementation (#80874)
rust-lang/rust: src/doc/rustdoc/src/linking-to-items-by-name.md

pub struct Bar;
```

> Unlike normal markdown, `[bar][Bar]` syntax is also supported without needing a

Reviewer: nit:

```suggestion
Unlike normal Markdown, `[bar][Bar]` syntax is also supported without needing a
```

> `[Bar]: ...` reference link, and links are case-sensitive.

Reviewer: We should probably note somewhere that non-intra-doc links are case-**insensitive**; see #80882.

Backticks around the link will be stripped, so ``[`Option`]`` will correctly
link to `Option`.

## Valid links

You can refer to anything in scope, and use paths, including `Self`, `self`, `super`, and
`crate`. Associated items (functions, types, and constants) are supported, but [not for blanket
trait implementations][#79682]. Rustdoc also supports linking to the following primitives, which
have no path and cannot be imported:

> - [`slice`](../std/primitive.slice.html)

Reviewer 2: I'd collapse this into a single sentence. Most people don't need to care about this.

> - [`slice`](../../std/primitive.slice.html)
> - [`array`](../../std/primitive.array.html)
> - [`tuple`](../../std/primitive.tuple.html)
> - [`unit`](../../std/primitive.unit.html)
> - [`fn`](../../std/primitive.fn.html)
> - [`pointer`](../../std/primitive.pointer.html), `*`, `*const`, or `*mut`
> - [`reference`](../../std/primitive.reference.html), `&`, or `&mut`
> - [`never`](../../std/primitive.never.html) or `!`

Reviewer: I think you need to remove one layer of `../`:

```
rustdoc/linking-to-items-by-name.html:205: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/std/primitive.slice.html
rustdoc/linking-to-items-by-name.html:206: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/std/primitive.array.html
rustdoc/linking-to-items-by-name.html:207: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/std/primitive.tuple.html
rustdoc/linking-to-items-by-name.html:208: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/std/primitive.unit.html
rustdoc/linking-to-items-by-name.html:209: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/std/primitive.fn.html
rustdoc/linking-to-items-by-name.html:210: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/std/primitive.pointer.html
rustdoc/linking-to-items-by-name.html:211: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/std/primitive.reference.html
rustdoc/linking-to-items-by-name.html:212: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/std/primitive.never.html
rustdoc/print.html:1309: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/std/primitive.slice.html
rustdoc/print.html:1310: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/std/primitive.array.html
rustdoc/print.html:1311: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/std/primitive.tuple.html
rustdoc/print.html:1312: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/std/primitive.unit.html
rustdoc/print.html:1313: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/std/primitive.fn.html
rustdoc/print.html:1314: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/std/primitive.pointer.html
rustdoc/print.html:1315: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/std/primitive.reference.html
rustdoc/print.html:1316: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/std/primitive.never.html
```

```suggestion
- [`slice`](../std/primitive.slice.html)
- [`array`](../std/primitive.array.html)
- [`tuple`](../std/primitive.tuple.html)
- [`unit`](../std/primitive.unit.html)
- [`fn`](../std/primitive.fn.html)
- [`pointer`](../std/primitive.pointer.html), `*`, `*const`, or `*mut`
- [`reference`](../std/primitive.reference.html), `&`, or `&mut`
- [`never`](../std/primitive.never.html) or `!`
```

> - [`never`](https://doc.rust-lang.org/std/primitive.never.html) or `!`

Reviewer: I think we're supposed to use relative links in docs so that they refer to the same release channel. But these types don't change much, so probably not a big deal.

Author: Done, thanks.

> 

Reviewer: You forgot `str`:

```suggestion
- [`str`](https://doc.rust-lang.org/std/primitive.str.html)

```

Reviewer: Ah, okay.

struct MySpecialFormatter;
```

## Namespaces and Disambiguators

> Paths in Rust have three namespaces: type, value, and macro. Item names must be unique within

Reviewer: It may not be necessary, but do you want to note that the type namespace includes modules?

> their namespace, but can overlap with items outside of their namespace. In case of ambiguity,

Reviewer: ```suggestion
their namespace, but can overlap with items in other namespaces. In case of ambiguity,
```

rustdoc will warn about the ambiguity and ask you to disambiguate, which can be done by using a
prefix like `struct@`, `enum@`, `type@`, `trait@`, `union@`, `const@`, `static@`, `value@`,

> `fn@`, `function@`, `mod@`, `module@`, `method@`, `prim@`, `primitive@`, `macro@`, or `derive@`:

Reviewer: It feels weird to through such a huge wall of text at people. Do you want to change this to be a list like

- `struct@`
- `enum@`
- `type@`

instead?

fn Foo() {}
```

> The following prefixes can be used:

Reviewer 2: I'd collapse this into a single sentence. Most people don't need to care about this, and rustdoc helps with diagnostics when they do.

- `struct@`
- `enum@`
- `type@`
- `trait@`
- `union@`
- `const@`
- `static@`
- `value@`
- `fn@` / `function@` / `method@`
- `mod@` / `module@`
- `prim@` / `primitive@`
- `macro@`
- `derive@`

These prefixes will be stripped when displayed in the documentation, so `[struct@Foo]`
will be rendered as `Foo`.

You can also disambiguate for functions by adding `()` after the function name,
or for macros by adding `!` after the macro name:

```rust
/// This is different from [`foo!`]
fn foo() {}

/// This is different from [`foo()`]
macro_rules! foo {
  () => {}
}
```

## Warnings, re-exports, and scoping

Links are resolved in the scope of the module where the item is defined, even
when the item is re-exported. If a link from another crate fails to resolve, no
warning is given.

When re-exporting an item, rustdoc allows adding additional documentation to it.

> That additional documentation will be resolved in scope of the re-export, not

Reviewer: You're missing "the" here:

```suggestion
That additional documentation will be resolved in the scope of the re-export, not
```

the original, allowing you to link to items in the new crate. The new links

> will still give a warning if they fail to resolve.

Reviewer: These two paragraph seem contradictory. The first one says that links are resolved in the module where the item is defined, even when it's re-exported, while the second says that documentation you add on a re-export will be resolved in the re-export crate. Which one is it? :P

Author: The *original* links will be resolved in the original scope. The *new* links will be resolved in the new scope. The idea is for this to work:

```rust
// crate 1
/// Link to [f()]
pub struct S;
pub fn f() {}

// crate2
/// Link to [g()]
pub use crate1::S;
pub fn g() {}

Reviewer: I think you should say something like that in the docs, because I couldn't tell what you meant from the way it is right now.

Author: I couldn't think of a way to explain this better, but I did add an example of both scenarios.

fn Foo() {}
```

## Warnings, re-exports, and scoping

> Links are resolved in the current module scope, even when re-exported. If a link from another

Reviewer: What does "even when re-exported" mean here? Can you elaborate?

crate fails to resolve, no warning is given.

> When re-exporting an item, rustdoc allows additional documentation to it. That documentation will

Reviewer: I think you're missing a word here.

```suggestion
When re-exporting an item, rustdoc allows adding additional documentation to it. That documentation will
```

Reviewer: We should probably define what re-exports are or link to the `doc(inline)` page somewhere.

> be resolved in the new scope, not the original, allowing you to link to items in the current
> crate. The new links will still give a warning if they fail to resolve.

Reviewer: I would say something along the lines of (wording could be improved):

```suggestion
be resolved in the scope of the re-export, not the original, allowing you to link to items in the crate that's re-exporting.
The new links will still give a warning if they fail to resolve.
```

to be extra-clear what "new" and "current" mean.

```rust
/// See also [foo()]
pub use std::process::Command;

pub fn foo() {}
```

> This is especially useful for proc-macros, which must always be in their own dedicated crate.

Reviewer: ```suggestion
This is especially useful for proc-macros, which must always be defined in their own dedicated crate.
```

Note: Because of how `macro_rules!` macros are scoped in Rust, the intra-doc links of a
`macro_rules!` macro will be resolved [relative to the crate root][#72243], as opposed to the
module it is defined in.

If links do not look 'sufficiently like' an intra-doc link, they will be ignored and no warning
will be given, even if the link fails to resolve. For example, any link containing `/` or `[]`
characters will be ignored. You can see the full criteria for 'sufficiently like' in [the source

> code].

Reviewer: I'm not sure if it's a good idea to link to the source code here. For one, it will very easily become out-of-date because it's tied to a particular commit. I think we should either skip this or include the list directly in the documentation.

Reviewer: > Maybe I could change the commit hash to `master` instead? That might have wrong line numbers though.

Yeah, I think that would be very confusing.

> Hmm, I kind of prefer not to tie this down so we can update it in the future.

What do you mean? Documenting it isn't "tying it down".

Reviewer 2: So, this is a not a reference, this is a _book_, and I don't think it should go into that much detail. We do not need to document every last corner of the implementation here. Let's just drop this section.
